File: elixir/lib/2024/11_biological_warfare/ec202411.ex
defmodule EverybodyCodes2024.Day11 do
  @moduledoc """
  Everybody Codes 2024, day 11: Biological Warfare
  """
  require EverybodyCodes

  @doc """
  Parse the input data
  """
  def parse(puzzle_input) do
    puzzle_input
    |> String.split("\n", trim: true)
    |> Enum.map(fn line ->
      [parent, children] = String.split(line, ":")
      {parent, children |> String.split(",") |> Enum.frequencies()}
    end)
    |> Enum.into(%{})
  end

  @doc """
  Solve part 1
  """
  def part1(puzzle_input) do
    puzzle_input |> parse() |> calculate_population(%{"A" => 1}, 4)
  end

  @doc """
  Solve part 2
  """
  def part2(puzzle_input) do
    puzzle_input |> parse() |> calculate_population(%{"Z" => 1}, 10)
  end

  @doc """
  Solve part 3
  """
  def part3(puzzle_input) do
    conversions = puzzle_input |> parse()

    conversions
    |> Map.keys()
    |> Enum.map(&calculate_population(conversions, %{&1 => 1}, 20))
    |> then(fn populations -> Enum.max(populations) - Enum.min(populations) end)
  end

  @doc """
  Calculate the population n days into the future.

  ## Example

      A -> BC -> AAB -> BCBCAA -> AABAABBCBC -> BCBCAABCBCAAAABAAB

      iex> conversions = %{"A" => %{"B" => 1, "C" => 1}, "B" => %{"A" => 2}, "C" => %{"B" => 1}}
      iex> calculate_population(conversions, %{"A" => 1}, 5)
      18
  """
  def calculate_population(_, population, 0), do: population |> Map.values() |> Enum.sum()

  def calculate_population(conversions, population, num_days) do
    calculate_population(
      conversions,
      Enum.reduce(population, %{}, fn {termite, count}, acc ->
        Enum.reduce(conversions[termite], acc, fn {child, child_count}, acc ->
          Map.update(acc, child, count * child_count, &(&1 + count * child_count))
        end)
      end),
      num_days - 1
    )
  end

  def main(path1, path2, path3) do
    EverybodyCodes.solve(&part1/1, &part2/1, &part3/1, path1, path2, path3)
  end
end

File: python/src/2024_the-kingdom-of-algorithmia/11_biological-warfare/ec202411.py
"""Everybody Codes 11, 2024: Biological Warfare"""

import collections


def parse(puzzle_input: str) -> dict[str, dict[str, int]]:
    """Parse the puzzle input.

    Flip the dictionary to map children to their parents to simplify
    calculations later.
    """
    conversions = collections.defaultdict(list)
    for line in puzzle_input.split("\n"):
        parent, _, children = line.partition(":")
        for child in children.split(","):
            conversions[child].append(parent)
    return {
        child: collections.Counter(parents) for child, parents in conversions.items()
    }


def part1(puzzle_input: str) -> int:
    """Solve part 1."""
    conversions = parse(puzzle_input)
    return calculate_population(conversions, initial="A", num_days=4)


def part2(puzzle_input: str) -> int:
    """Solve part 2."""
    conversions = parse(puzzle_input)
    return calculate_population(conversions, initial="Z", num_days=10)


def part3(puzzle_input: str) -> int:
    """Solve part 3."""
    conversions = parse(puzzle_input)
    pop_sizes = [
        calculate_population(conversions, initial=termite, num_days=20)
        for termite in conversions.keys()
    ]
    return max(pop_sizes) - min(pop_sizes)


def grow_population(
    conversions: dict[str, dict[str, int]], population: dict[str, int]
) -> dict[str, int]:
    """Grow the population one day

    ## Example

    >>> conversions = {"B": {"A": 1}, "C": {"A": 1, "B": 1}, "A": {"B": 1, "C": 1}}
    >>> grow_population(conversions, {"B":1, "C": 1})
    {'B': 0, 'C': 1, 'A': 2}
    """
    return {
        termite: sum(
            population.get(parent, 0) * count for parent, count in parents.items()
        )
        for termite, parents in conversions.items()
    }


def calculate_population(
    conversions: dict[str, dict[str, int]], initial: str, num_days: int
) -> int:
    """Calculate the population after the given number of days.

    ## Example

        C -> A -> BC -> CAA -> ABCBC

    >>> conversions = {"B": {"A": 1}, "C": {"A": 1, "B": 1}, "A": {"B": 1, "C": 1}}
    >>> calculate_population(conversions, initial="C", num_days=4)
    5
    """
    population = {initial: 1}
    for _ in range(num_days):
        population = grow_population(conversions, population)
    return sum(population.values())
